Kernel/FileSystem/Inode.h

```diff
@@ -134,6 +137,7 @@ class Inode : public RefCounted<Inode>
     RefPtr<LocalSocket> m_socket;
     HashTable<InodeWatcher*> m_watchers;
     bool m_metadata_dirty { false };
+    RefPtr<FIFO> m_fifo;
```

Comment:
So, what keeps the inode alive?

Say I `mkfifo("/foo")`, and that creates an inode and sets its `m_fifo` to a new FIFO object. Then the inode gets dropped and destructed, and so does the FIFO object since nothing else references it.

Now I `open("/foo")`, and that should open that same FIFO, but instead it finds me another inode (which is supposed to be "the same" FS inode, but it's a different object in memory), which gives me a fresh FIFO.

Reply:
It is expected behavior that the FIFO would disappear when no one has it open. To keep it alive when it is open, I set the original inode of the `FileDescription` to the inode (after this review).

Reply:
> It is expected behavior that the FIFO would disappear when no one has it open.

I still don't see how that could possibly work. Someone calls `mkfifo()`, a new FIFO is created, and (since no one has opened it yet) immediately destroyed?

Reply:
When someone calls mkfifo, a fifo special file is created. `m_fifo` is actually a pipe, due to a weird naming choice in serenity. the code to actually create a fifo file already existed, it just wouldn't be opened properly. The pipe `m_fifo` is created when the inode is opened.

from the Linux Man pages, `FIFO(7)`
> The kernel maintains exactly one pipe object for each FIFO special file that is opened by at least one process.

Reply:
Ah, I see, so `mkfifo()` doesn't actually create the FIFO, only a record in the FS about one.

Yeah, `set_original_inode()` would probably do the trick. I wonder how this works for local sockets, and whether it does actually work. (TmpFS, of course, never drops otherwise unused inodes, so that might be alleviating the problem.)